### PR TITLE
fix(transformerblock): conditionally initialize cross attention compo…

### PR DESCRIPTION
**Pull request by yunhaoli24**
Transformerblock

This PR improves the initialization logic of TransformerBlock by adding conditional layer creation for cross-attention.
Previously, the cross-attention components (norm_cross_attn and cross_attn) were always initialized, even when with_cross_attention=False. This introduced unnecessary memory overhead and additional parameters that were never used during forward passes.

With this change, the cross-attention layers are initialized only when with_cross_attention=True.
This ensures cleaner module definitions, reduced memory usage, and avoids confusion about unused layers in models that do not require cross-attention.

Fixes # .

### Description

Before
```python
...
class TransformerBlock(nn.Module):


    def __init__(
        self,
...
    ) -> None:
       ...
        self.norm2 = nn.LayerNorm(hidden_size)
        self.with_cross_attention = with_cross_attention

        self.norm_cross_attn = nn.LayerNorm(hidden_size)
        self.cross_attn = CrossAttentionBlock(
            hidden_size=hidden_size,
            num_heads=num_heads,
            dropout_rate=dropout_rate,
            qkv_bias=qkv_bias,
            causal=False,
            use_flash_attention=use_flash_attention,
        )
```

After
```python
...
class TransformerBlock(nn.Module):


    def __init__(
        self,
...
    ) -> None:
       ...
        self.norm2 = nn.LayerNorm(hidden_size)
        self.with_cross_attention = with_cross_attention

        if with_cross_attention:
            self.norm_cross_attn = nn.LayerNorm(hidden_size)
            self.cross_attn = CrossAttentionBlock(
                hidden_size=hidden_size,
                num_heads=num_heads,
                dropout_rate=dropout_rate,
                qkv_bias=qkv_bias,
                causal=False,
                use_flash_attention=use_flash_attention,
            )
```
### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by ericspod** [comment]
Hi @yunhaoli24 thanks for the contribution, but as the Coderabbit comment notes this will break existing stored weight files since the members will not always exist. Your change is more efficient but I think requires more consideration to avoid this issue, and also should be documented in the docstring that this is happening. We are working on a quick release to be done shortly but then we can return to this PR.

**Comment by yunhaoli24** [comment]
Thanks for the feedback! I agree with your point about checkpoint compatibility; that’s indeed an important concern.
At the same time, I’d like to note that in a DDP multi-GPU setting, initializing but not using the cross_attn layers can lead to torch errors about unused parameters during training. So this change could also help avoid those issues in addition to saving memory.
I think a good compromise might be to add backward-compatibility handling when loading checkpoints, or at least documenting this behavior clearly in the docstring.

**Comment by ericspod** [comment]
> Thanks for the feedback! I agree with your point about checkpoint compatibility; that’s indeed an important concern. At the same time, I’d like to note that in a DDP multi-GPU setting, initializing but not using the cross_attn layers can lead to torch errors about unused parameters during training. So this change could also help avoid those issues in addition to saving memory. I think a good compromise might be to add backward-compatibility handling when loading checkpoints, or at least documenting this behavior clearly in the docstring.

Yes I have DDP errors related to this so you're right, this is one advantage to consider. Please do update the docstring to record what's going on here. One other thing to consider is Torchscript compatibility which I can't say is being tested thoroughly enough for this class. I know it has in the past complained about absent members even if the Python code works, you may need to always define the members but use `nn.Identity` as their values when `with_cross_attention` is False (in which case the condition check isn't needed in `forward` either).

**Comment by yunhaoli24** [comment]
Thanks for the suggestion! Regarding defining cross_attn as nn.Identity, I have a few concerns:

* Interface mismatch: The current forward unconditionally calls self.cross_attn(self.norm_cross_attn(x), context=context), whereas nn.Identity.forward only accepts a single input and doesn’t support a context positional or keyword argument. This would raise an “unexpected keyword argument 'context'” error in both Eager and TorchScript.
* Semantic change: Even if we change the call to pass only x to avoid the error, nn.Identity returns the normalized x as-is. That adds a non-zero residual when with_cross_attention=False, which breaks the expectation that disabling cross-attn should not alter the output.
* TorchScript compatibility: Scripting performs static signature checks. Invoking a module with a context argument when the target doesn’t accept it will fail at script/trace time or produce an inconsistent graph.

**Comment by ericspod** [comment]
> Thanks for the suggestion! Regarding defining cross_attn as nn.Identity, I have a few concerns:

Sorry for the delay in replying. I see the issue you're mentioning about `nn.Identity`, one possible solution is to define a new `nn.Module` subclass whose `forward` method accepts two arguments but returns 0 always, it's less computationally efficient however. We should double check that Torchscript works with your change as it is first and that the unit tests are thorough enough in that regard, but a second solution is to define `self.cross_attn` as `nn.Identity` in the constructor (and always define `self.norm_cross_attn`) when `with_cross_attention` is false, and just not use it in the `forward` method. 

@virginiafdez could you please assist here as well? Thanks!

**Comment by ericspod** [comment]
I would like to mention that another PR https://github.com/Project-MONAI/MONAI/pull/8584 is looking at the same problem. I suggested we consolidate efforts here.

**Comment by yunhaoli24** [comment]
I’d be happy to continue working on this PR. Please let me know if there’s anything that needs to be modified or improved. @ericspod

**Comment by virginiafdez** [comment]
Hi @yunhaoli24, many thanks for spotting this and apologies for the delay. Since this issue is duplicated and also addressed in the PR @ericspod linked, which also addresses the checkpoint problem, we will move the fix to that PR for the time being. Further tests and fixes are required, but I will keep these two PRs open for the time being and make sure I update as soon as the fix's done.
